src/crimson/osd/osd_operations/pg_advance_map.cc

```diff
@@ -133,4 +125,76 @@ seastar::future<> PGAdvanceMap::start()
   });
 }
 
+seastar::future<> PGAdvanceMap::split_pg(
+    std::set<std::pair<spg_t, epoch_t>> split_children_info,
+    cached_map_t next_map)
+{
+  using cached_map_t = OSDMapService::cached_map_t;
+  logger().debug("{}: start", *this);
+
+  for (auto& child_pg_info : split_children_info) {
+    auto child_pgid = child_pg_info.first;
+    auto pg_epoch = child_pg_info.second;
+    children_pgids.insert(child_pgid);
+
+    // Map each child pg ID to a core
+    auto core = co_await shard_services.create_split_pg_mapping(child_pgid, seastar::this_shard_id());
+    logger().debug(" PG {} mapped to {}", child_pgid.pgid, core);
+    logger().debug(" {} map epoch: {}", child_pgid.pgid, pg_epoch);
+    cached_map_t map = co_await shard_services.get_map(pg_epoch);
+    auto child_pg = co_await shard_services.make_pg(std::move(map), child_pgid, true);
+
+    logger().debug(" Parent PG: {}", pg->get_pgid());
+    logger().debug(" Child PG ID: {}", child_pg->get_pgid());
+    unsigned new_pg_num = next_map->get_pg_num(pg->get_pgid().pool());
+    // Depending on the new_pg_num the parent PG's collection is split.
+    // The child PG will be initiated with this split collection.
+    unsigned split_bits = child_pg->get_pgid().get_split_bits(new_pg_num);
+    logger().debug(" pg num is {}, m_seed is {}, split bits is {}",
+	            new_pg_num, child_pg->get_pgid().ps(), split_bits);
+
+    co_await pg->split_colls(child_pg->get_pgid(), split_bits, child_pg->get_pgid().ps(),
+                             &child_pg->get_pgpool().info, rctx.transaction);
+    logger().debug(" {} split collection done", child_pg->get_pgid());
+    // Update the child PG's info from the parent PG
+    pg->split_into(child_pg->get_pgid().pgid, child_pg, split_bits);
+
+    co_await this->template with_blocking_event<PGMap::PGCreationBlockingEvent>(
+	[this, child_pg, next_map] (auto&& trigger) -> seastar::future<> {
```

Comment:
Mixing lambdas and coroutines can be tricky: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rcoro-capture

Reply:
Done, moved it to another function

Reply:
This is still a capturing lambda used by a coroutine, we would need to move the lambda _into_ the function to avoid this.

Reply:
Either wrapping the lambda with `seastar::coroutine::lambda` could work:
```
co_await this->template with_blocking_event<PGMap::PGCreationBlockingEvent>(
  seastar::coroutine::lambda([this, child_pg, next_map] (auto&& trigger) -> seastar::future<> {
    return handle_split_pg_creation(child_pg, next_map, std::move(trigger));
  }));
```